index.js

```diff
@@ -78,36 +60,41 @@ class AppCacheCompiler {
     }
   }
 
-  _format(obj) {
-    return (Array.from(Object.keys(obj).sort()).map(k => `${k} ${obj[k]}`)).join('\n');
-  };
+  _compile(file) {
+    try {
+      this._processFile(file);
+      return Promise.resolve(file);
+    } catch (err) {
+      return Promise.reject(err);
+    }
+  }
 
-  _readStream(path) {
-    return new Promise((resolve, reject) => {
-      try {
-        let shasum = crypto.createHash('sha1');
-        const stream = fs.createReadStream(path);
-
-        stream.on('data', data => shasum.update(data));
-
-        stream.on('end', () => {
-          this.shasums.push(shasum.digest('hex'));
-          if (this.shasums.length === this.paths.length) {
-            shasum = crypto.createHash('sha1');
-            shasum.update(this.shasums.sort().join(), 'ascii');
-            this.changedFileContentShasum = shasum.digest('hex');
-          }
-          resolve();
-        });
-        stream.on('finish', resolve);
-        stream.on('error', reject);
-
-      } catch (error) {
-        reject(error);
-      }
-    });
+  _processFile(file) {
+    const path = file.path;
+    let shasum = crypto.createHash('sha1');
+    shasum.update(file.data);
+    this.shasums.push(shasum.digest('hex'));
+
+    if (this._includePath(path, this.paths)) {
+      this.paths.push(path);
+      this.paths.sort();
+    }
+
+    if (this.shasums.length === this.paths.length) {
+      shasum = crypto.createHash('sha1');
+      shasum.update(this.shasums.sort().join(), 'ascii');
+      this.changedFileContentShasum = shasum.digest('hex');
+    }
+  }
+
+  _includePath(path, paths) {
+    return !path.endsWith('.appcache') && !this.config.ignore.test(path) && !paths.includes(path);
   }
 
+  _format(obj) {
+    return (Array.from(Object.keys(obj).sort(), k => `${k} ${obj[k]}`)).join('\n');
```

Comment:
Didn't we agree this should not be method?

Uses `this` => method
Otherwise => helper

Easy.